experimental/packages/otlp-transformer/src/trace/index.ts

```diff
@@ -0,0 +1,93 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import type { InstrumentationLibrary } from '@opentelemetry/core';
+import type { Resource } from '@opentelemetry/resources';
+import type { ReadableSpan } from '@opentelemetry/sdk-trace-base';
+import { toAttributes } from '../common/internal';
+import { toSpan } from './internal';
+import { IExportTraceServiceRequest } from './types';
+
+export function createExportTraceServiceRequest(spans: ReadableSpan[]): IExportTraceServiceRequest | null {
+  if (spans.length === 0) {
+    return null;
+  }
+
+  return {
+    resourceSpans: groupByResource(spans).map(({ resource, resourceSpans, schemaUrl: resourceSchemaUrl }) => ({
+      resource: {
+        attributes: toAttributes(resource.attributes),
+        droppedAttributesCount: 0,
+      },
+      instrumentationLibrarySpans: resourceSpans.map(({ instrumentationLibrary, instrumentationLibrarySpans, schemaUrl: librarySchemaUrl }) => ({
+        instrumentationLibrary: {
+          name: instrumentationLibrary.name,
+          version: instrumentationLibrary.version,
+        },
+        spans: instrumentationLibrarySpans.map(toSpan),
+        schemaUrl: librarySchemaUrl,
+      })),
+      schemaUrl: resourceSchemaUrl,
+    }))
+  };
+}
+
+type ResourceSpans = {
+  resource: Resource,
+  resourceSpans: InstrumentationLibrarySpans[],
+  schemaUrl?: string,
+};
+
+type InstrumentationLibrarySpans = {
+  instrumentationLibrary: InstrumentationLibrary,
+  instrumentationLibrarySpans: ReadableSpan[],
+  schemaUrl?: string,
+};
+
+function groupByResource(SpanRecords: ReadableSpan[]): ResourceSpans[] {
+  const resourceMap: Map<Resource, Map<string, ReadableSpan[]>> = new Map();
+
+  for (const record of SpanRecords) {
+    let ilmMap = resourceMap.get(record.resource);
+
+    if (!ilmMap) {
+      ilmMap = new Map();
+      resourceMap.set(record.resource, ilmMap);
+    }
+
+    const instrumentationLibraryKey = JSON.stringify([record.instrumentationLibrary.name, record.instrumentationLibrary.version, record.instrumentationLibrary.schemaUrl]);
+    let records = ilmMap.get(instrumentationLibraryKey);
+
+    if (!records) {
+      records = [];
+      ilmMap.set(instrumentationLibraryKey, records);
+    }
+
+    records.push(record);
+  }
+
+  const out: ResourceSpans[] = [];
+
+  for (const [resource, ilmMap] of resourceMap.entries()) {
+    const resourceSpans: InstrumentationLibrarySpans[] = [];
+    for (const [instrumentationLibraryKey, instrumentationLibrarySpans] of ilmMap.entries()) {
+      const [name, version, schemaUrl] = JSON.parse(instrumentationLibraryKey);
+      resourceSpans.push({ instrumentationLibrary: { name, version }, instrumentationLibrarySpans, schemaUrl });
+    }
+    out.push({ resource, resourceSpans });
+  }
+
+  return out;
```

Comment:
I think this can be simplified to:
```js
return resourceMap.entries().map( ([resource, ilmMap]) => ({
  resource,
  resourceSpans: ilmMap.entries().map( ([instrumentationLibraryKey, instrumentationLibrarySpans] => {
      const [name, version, schemaUrl] = JSON.parse(instrumentationLibraryKey);
      return { instrumentationLibrary: { name, version }, instrumentationLibrarySpans, schemaUrl };    
  },
})
```

It's a matter of style, so feel free to ignore it if you don't like. I usually prefer this "functional" writing over creating a side effect inside the loop on some external variable.

Reply:
You can't map on map `entries()` iterable so it would require converting it to an array. Usually I would agree with you but in this case I think the functional version is a little harder to read and would require unnecessary allocations.

Reply:
Right, I missed that.
Thought it behaves like `Object. entries()` 😅
I vaguely remember that `push`ing into an array creates many reallocations as more elements are added, while `map` can preallocate the array with the exact size thus is preferred. But I guess it's even worst to transform the `Map` to `array` in this case.